### PR TITLE
Optimize MTP distillation loss with boundary-token P2P exchange

### PR DESCRIPTION
**Pull request by lshpku**
### PR Category
Communication Library

### PR Types
Performance

### Description
实现了 MTP Distillation Loss 的无冗余通信的 CP

原来的实现是先做 cp_group 内的 gather，然后 shift，再 scatter，这个 gather 会产生一个未 cp 切分的中间变量，失去了 cp 的意义，长文下很容易 OOM；本 PR 则只交换 shift 的边界 token，不会产生中间变量，也不会冗余发送

优化后的分支仅对 cp_balance_mode == "contiguous_allgather" 生效，原来的 dualchunk 布局依然支持，但是就没有优化，可能会 OOM

由于本 PR 只是对数据搬运的优化，不涉及 reduce 等，因此无任何精度 diff，单测保证 atol=0 rtol=0

### 是否引起精度变化
否

### CODE REVIEW

File: src/paddlefleet/context_parallel_utils.py [hunk 1]
```diff
@@ -1947,3 +1947,113 @@ def flashmask_attention_cp(
...
+
+        return _mtp_distillation_loss_shift_forward(
+            tensor, num_nextn_predict_layers, group
+        )
+
+    @staticmethod
+    def backward(ctx, output_grad):
+        output_grad = output_grad.reshape(
+            [ctx.batch_size, ctx.seq_len + ctx.nextn - 1, ctx.hidden_size]
+        )
```
Comment (risemeup1111) [inline src/paddlefleet/context_parallel_utils.py:2060]: P1 **优先级：P1**

`forward` 只保存了 `ctx.num_nextn_predict_layers`，但 `backward` 这里读取的是不存在的 `ctx.nextn`。只要 MTP distillation loss 参与梯度，CP>1 路径会在进入 `_mtp_distillation_loss_shift_backward` 前抛 `AttributeError`，本 PR 声明的反向精度对齐也无法执行。请使用已保存的 `ctx.num_nextn_predict_layers`。

```suggestion
        output_grad = output_grad.reshape(
            [
                ctx.batch_size,
                ctx.seq_len + ctx.num_nextn_predict_layers - 1,
                ctx.hidden_size,
            ]
        )
```

File: src/paddlefleet/context_parallel_utils.py [hunk 2]
```diff
@@ -1947,3 +1947,113 @@ def flashmask_attention_cp(
...
+        recv_window = paddle.empty([bs, nextn, hidden], tensor.dtype)
+        ops.append(dist.P2POp(dist.irecv, recv_window, recv_rank, group))
+    else:
+        recv_window = paddle.zeros([bs, nextn, hidden], tensor.dtype)
+
+    _wait_all(dist.batch_isend_irecv(ops))
+
+    # output has shape [bs, seq_len, hidden]
+    output = paddle.nn.functional.pad(
+        tensor[:, :-nextn], [0, 0, 1], mode="constant", value=0
```
Comment (risemeup1111) [inline src/paddlefleet/context_parallel_utils.py:2003]: P1 **优先级：P1**

这里的 `pad` 参数长度是 3，按 Paddle 的 pad 约定需要成对描述每个维度的 before/after padding；仓库里其它 3D 张量补序列维也使用 4 个值，例如 `[0, 0, 0, pad_size]`。修复 `ctx.nextn` 后，当前写法仍会让反向在这里因非法 pad 配置失败。这里想在 `[B, S-nextn, H]` 的序列维左侧补 1，应改为：

```suggestion
        tensor[:, :-nextn], [0, 0, 1, 0], mode="constant", value=0
```

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 3]
```diff
@@ -557,9 +558,9 @@ def forward(self, logits: Tensor | list, labels: Tensor) -> Tensor:
                 else:
                     target_p_self_op_dist = nn.Softmax(axis=2)(logits[0])
                 if get_context_parallel_world_size() > 1:
-                    target_p_self_op_dist = ContextParallelGatherOp.apply(
+                    target_p_self_op_dist = MTPDistillationLossShift.apply(
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/language_loss/language_loss.py]: P1 **优先级：P1**

这里把通用 `cp_balance_mode` 直接传给 `MTPDistillationLossShift`，但新增 PyLayer 在 `forward` 里只允许 `contiguous_allgather`。仓库默认值仍是 `dualchunk_allgather`，而 `TransformerConfig` 也允许 `dualchunk_allgather`/`contiguous_a2a`，所以默认 CP + `mtp_distillation_loss=True` 会在前向阶段直接 assert，无法训练。请在调用前按模式分流：`contiguous_allgather` 走新的 P2P shift，其它合法模式保留原来的 gather/slice/scatter 路径，或在配置校验阶段显式禁止这些组合并给出清晰错误。示例形状：

```python
if get_context_parallel_world_size() > 1:
    if self.config.cp_balance_mode == "contiguous_allgather":
        target_p_self_op_dist = MTPDistillationLossShift.apply(
            target_p_self_op_dist,
            self.config.num_nextn_predict_layers,
            mode=self.config.cp_balance_mode,
        )
    else:
        target_p_self_op_dist = ContextParallelGatherOp.apply(
            target_p_self_op_dist, axis=1, mode=self.config.cp_balance_mode
        )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现了会阻塞 CP + MTP distillation 训练路径的运行时问题，具体位置和建议已放在行级评论里。CI 目前还有多项任务在运行中，建议修复后再结合多卡用例复查。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/context_parallel_utils.py (hunk 2 above)
Comment (liuhao2638) [reply src/paddlefleet/context_parallel_utils.py:2003]: P1 **优先级：P1**

补充一条新提交后的证据：这次新增的 `test_forward_backward()` 会在 `loss_tgt.backward()` 进入 `MTPDistillationLossShift.backward()`，因此修复测试启动配置后仍会命中这里的 `pad(..., [0, 0, 1])`。建议同时把该行改为 `[0, 0, 1, 0]`，否则新增反向 parity 测试会被同一个非法 pad 参数阻塞。

File: tests/multi_card_tests/tensor_parallel/test_mtp_distillation_loss_cp.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class TestMTPDistillationLossCP(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        Utils.initialize_model_parallel(
+            context_parallel_size=8,
+            expert_parallel_size=8,
+            sharding_parallel_size=8,
+        )
```
Comment (liuhao2638) [inline tests/multi_card_tests/tensor_parallel/test_mtp_distillation_loss_cp.py:48]: P1 **优先级：P1**

这个测试文件位于 `tests/multi_card_tests/tensor_parallel/`，`tests/test_configs.yaml` 里 `tests/multi_card_tests/tensor_parallel/*.py` 会优先匹配并用 4 张卡启动；但这里初始化要求 `context_parallel_size=8`、`expert_parallel_size=8`、`sharding_parallel_size=8`。在 4 卡任务里这些 degree 超过实际 world size，新增测试会在 fleet 初始化阶段失败，无法提供本 PR 需要的反向覆盖。请改成与该目录 4 卡配置一致的 CP 测试，或把该用例移到 8 卡配置匹配的位置/在 YAML 中添加更靠前的特例。当前用例不涉及 MoE，直接用 4 卡 CP 即可：

```suggestion
        Utils.initialize_model_parallel(
            context_parallel_size=4,
            sharding_parallel_size=4,
        )
```

**Review by liuhao2638 — CHANGES_REQUESTED** [review]
复查了新提交。`ctx.nextn` 和 `cp_balance_mode` 分流问题看起来已经处理；但反向 `pad` 参数问题仍在，且新增多卡测试的启动 GPU 配置也会让测试在当前 CI 配置下初始化失败。具体建议已放在行级评论和原线程回复里。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: tests/multi_card_tests/tensor_parallel/test_mtp_distillation_loss_cp.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
```
Comment (risemeup1111) [inline tests/multi_card_tests/tensor_parallel/test_mtp_distillation_loss_cp.py:48]: P1 **优先级：P1**

这次提交只补了 `sys.path`，但这个用例仍在 `tests/multi_card_tests/tensor_parallel/` 下。`tests/test_configs.yaml` 的首个匹配规则会把 `tests/multi_card_tests/tensor_parallel/*.py` 用 4 张卡启动，而这里仍要求 `context_parallel_size=8`、`expert_parallel_size=8`、`sharding_parallel_size=8`，degree 超过实际 world size 后会在 fleet 初始化阶段失败，新增测试也就无法覆盖本 PR 的反向路径。当前用例不涉及 MoE，建议改成该目录匹配的 4 卡 CP 配置，或在 `tests/test_configs.yaml` 为此文件添加更靠前的 8 卡特例。

```suggestion
        Utils.initialize_model_parallel(
            context_parallel_size=4,
            sharding_parallel_size=4,
        )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
复查了新提交。新增的 `sys.path` 处理看起来是在修复测试导入问题，但多卡测试的启动卡数配置仍与用例初始化不匹配：该文件会按 4 卡运行，却仍初始化 8 卡 CP/EP/Sharding。具体建议已放在新的行级评论里。此前 `ctx.nextn` 和 `cp_balance_mode` 分流问题保持已修复状态。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: tests/multi_card_tests/tensor_parallel/test_mtp_distillation_loss_cp.py (hunk 5 above)
Comment (risemeup1111) [reply tests/multi_card_tests/tensor_parallel/test_mtp_distillation_loss_cp.py:48]: 这次提交把该测试的初始化配置改成了 4 卡 CI 规则下可用的 `world_size = 4`，并与同类 CP 测试的 hybrid degree 写法保持一致；这个测试启动卡数不匹配的问题我这边看已解决，后续以当前 CI 实跑结果为准。

**Review by risemeup1111 — APPROVED** [review]
复查了新提交：此前指出的多卡测试启动配置问题已经调整为与当前 4 卡测试规则匹配，我这边未发现新的需要阻塞合入的问题。CI 仍在运行中，建议以后续实跑结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by ForFishes — APPROVED** [review]
LGTM

**Comment by risemeup1111** [comment]
❌ Cherry-pick failed: Target branch `release/0.3` does not exist.

**Comment by risemeup1111** [comment]
✅ Cherry-pick successful! Created PR: https://github.com/PaddlePaddle/PaddleFleet/pull/1416
